src/cr/cube/matrix.py

```diff
@@ -2,157 +2,318 @@
+
+    @lazyproperty
+    def columns_base(self):
+        """1D/2D np.int64 ndarray of unweighted-N for each slice column/cell."""
+        # --- an MR_X slice produces a 2D column-base (each cell has its own N) ---
+        if self._rows_dimension.dimension_type == DT.MR_SUBVAR:
+            return self._assemble_matrix(
+                _SumSubtotals.blocks(self._cube_result_matrix, "columns_base")
+            )
+
+        # --- otherwise columns-base is a vector ---
+        return self._assemble_vector(
+            self._cube_result_matrix.columns_base,
+            self._column_subtotals,
+            self._column_order,
+        )
 
     @lazyproperty
-    def columns(self):
-        """A sequence of `_VectorAfterHiding` objects each representing a matrix column.
+    def columns_dimension_numeric_values(self):
+        """1D optional np.int/float64 ndarray of numeric-value for each column element.
 
-        Each column has all transformations applied, in particular, renaming, ordering,
-        insertions (subtotals) and hiding (including pruning) of elements. The value in
-        any given "cell" of each column also appears in a "cell" of its opposing row
-        vector. Accessing values row-wise or column-wise produces the same overall
-        matrix of values, although most commonly values are accessed row-wise.
+        A value of np.nan appears for a column element without a numeric-value. All
+        subtotal columns have a value of np.nan (subtotals have no numeric value).
         """
-        return tuple(
-            _VectorAfterHiding(column, self._assembled_rows)
-            for column in self._assembled_columns
-            if not column.hidden
+        elements = self._columns_dimension.valid_elements
+        return np.array(
+            [
+                (elements[idx].numeric_value if idx >= 0 else np.nan)
+                for idx in self._column_order
+            ]
         )
 
     @lazyproperty
-    def overlaps_tstats(self):
-        return (
-            self._unordered_matrix.overlaps_tstats
-            if self._unordered_matrix.is_augmented
-            else None
+    def columns_margin(self):
+        """1D/2D np.float64 ndarray of weighted-N for each column of this slice."""
+        # --- an MR_X slice produces a 2D columns-margin (each cell has its own N) ---
+        if self._rows_dimension.dimension_type == DT.MR_SUBVAR:
+            return self._assemble_matrix(
+                _SumSubtotals.blocks(self._cube_result_matrix, "columns_margin")
+            )
+
+        # --- otherwise columns-base is a vector ---
+        return self._assemble_vector(
+            self._cube_result_matrix.columns_margin,
+            self._column_subtotals,
+            self._column_order,
+        )
+
+    @lazyproperty
+    def inserted_column_idxs(self):
+        """tuple of int index of each subtotal column in slice."""
+        return tuple(i for i, col_idx in enumerate(self._column_order) if col_idx < 0)
```

Comment:
I'd like to see a comment here to explain why you're considering only the negative col_idx

Reply:
done